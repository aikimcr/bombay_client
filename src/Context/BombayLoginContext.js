import { createContext } from "react";

const BombayLoginContext = createContext({
  loggedIn: false,
  showLoginForm: false,
  setLoggedIn: null,
  setShowLogin: null,
});

export default BombayLoginContext;
