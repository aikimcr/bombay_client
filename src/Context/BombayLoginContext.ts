import { createContext, Dispatch } from "react";

interface BombayLoginContextValues {
  loggedIn: boolean;
  setLoggedIn?: Dispatch<boolean>;
}

export const BombayLoginContext = createContext({
  loggedIn: false,
  setLoggedIn: null,
});

export default BombayLoginContext;
