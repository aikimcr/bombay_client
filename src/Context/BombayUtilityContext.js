import { createContext } from 'react';

const BombayUtilityContext = createContext({});

export default BombayUtilityContext;
