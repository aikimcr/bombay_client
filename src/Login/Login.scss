@import '../global.scss';

.login-header {
    text-align: center;
}

.login-container {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: black;
}

.login-form {
    width: 600px;
    height: 400px;
    position: relative;
    top: calc(50% - 200px);
    left: calc(50% - 300px);
    background: white;
    padding-top: 60px;

    .info>div {
        width: 400px;
        display: grid;
        grid-template-columns: 100px 300px;
    }

    .info,
    .error,
    .controls {
        position: relative;
        left: 100px;
        width: 400px;
        margin-top: 20px;
    }

    .error {
        border: thin solid red;
        color: red;
        font-weight: bold;
        padding: 10px;
    }

    .controls {
        display: flex;
        justify-content: space-between;
    }

}

// Crossed out eye "\f070"; "\01F441";
// Eye "\f06e";