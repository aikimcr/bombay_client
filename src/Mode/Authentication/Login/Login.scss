@import '../../../global.scss';

.login-header {
    text-align: center;
}

.login-container {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
    width: 100%;
    height: 100%;
    background: transparent;
}

.login-form {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: start;
    gap: 1rem;
    width: 100%;
    height: 100%;
    background: transparent;
    @include form-container;

    .info {
        display: flex;
        flex-direction: column;
        justify-content: flex-start;
        align-items: flex-start;
        gap: 1rem;
        width: 100%;
        height: 100%;
    }

    .error {
        border: thin solid red;
        color: red;
        font-weight: bold;
        padding: 10px;
    }

    .controls {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: flex-start;
        flex-wrap: nowrap;
        gap: 1rem;
        width: 100%;
        height: 100%;
    }

}

// Crossed out eye "\f070"; "\01F441";
// Eye "\f06e";