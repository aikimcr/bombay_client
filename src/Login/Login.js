import { useEffect, useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';

import './Login.scss';

import BombayLoginContext from '../Context/BombayLoginContext';
import BombayModeContext from '../Context/BombayModeContext';
import BombayUtilityContext from '../Context/BombayUtilityContext';

import { login } from "../Network/Login";
import LabeledInput from '../Widgets/LabeledInput';

function Login(props) {
    const navigate = useNavigate();

    let timerHandle = null;

    const [error, setError] = useState(null);

    const loggedIn = useContext(BombayLoginContext);
    const mode = useContext(BombayModeContext);
    const { checkLoginState } = useContext(BombayUtilityContext);

    useEffect(() => {
        if (loggedIn) {
            navigate(`/${mode}`);
        }
    }, [loggedIn, mode, navigate]);

    function getCredentials() {
        // TODO: THere has to be a better way to do what these two lines are doing.
        const usernameInput = document.querySelector('[data-fieldname="username"] input');
        const passwordInput = document.querySelector('[data-fieldname="password"] input');

        return { username: usernameInput.value, password: passwordInput.value };
    }

    function clearTimer() {
        if (timerHandle) {
            clearTimeout(timerHandle);
            timerHandle = null;
        }
    }

    function handleChange(event) {
        timerHandle = setTimeout(() => {
            const loginButton = document.querySelector('.btn.login');

            clearTimer();

            const { username, password } = getCredentials();

            if (username.length > 0 && password.length > 0) {
                loginButton.classList.remove('disabled');
            } else {
                loginButton.classList.add('disabled');
            }
        }, 250);
    }

    function clearAllFields(event) {
        const loginButton = document.querySelector('.btn.login');

        clearTimer();

        document.querySelector('[data-fieldname="username"] input').value = '';
        document.querySelector('[data-fieldname="password"] input').value = '';
        loginButton.classList.add('disabled');
        setError(null);
    }

    async function doLogin(event) {
        const { username, password } = getCredentials();

        clearTimer();

        await login(username, password)
            .catch(err => {
                if (err.status === 401) {
                    setError('Username or password is incorrect.');
                } else {
                    setError(`${err.status}: ${err.message}`);
                }
            });

        checkLoginState();
    }

    if (loggedIn) {
        return '';
    } else {
        return (
            <div className="login-container">
                <div className="login-form">
                    <h1 className="login-header">Please Log In</h1>
                    <div className="info">
                        <LabeledInput
                            modelName='login'
                            fieldName='username'
                            labelText='User Name'
                            onChange={handleChange}
                        />
                        <LabeledInput
                            modelName='login'
                            fieldName='password'
                            labelText='Password'
                            type='password'
                            onChange={handleChange}
                        />
                    </div>
                    {error && <div className="error">{error}</div>}
                    <div className="controls">
                        <div className="clear btn" onClick={clearAllFields}>Clear All Fields</div>
                        <div className="login btn disabled" onClick={doLogin}>Login</div>
                    </div>
                </div>
            </div>
        );
    }
}

export default Login;