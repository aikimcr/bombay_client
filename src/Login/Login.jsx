import { useContext, useState } from "react";
import { createPortal } from "react-dom";

import "./Login.scss";

import BombayLoginContext from "../Context/BombayLoginContext";

import { loginStatus, login } from "../Network/Login";
import OldLabeledInput from "../Components/Widgets/OldLabeledInput";
import Button from "../Components/Widgets/Button";

function Login(props) {
  let timerHandle = null;

  const [error, setError] = useState(null);
  const [submitDisabled, setSubmitDisabled] = useState(true);

  const { loggedIn, setLoggedIn, showLoginForm } =
    useContext(BombayLoginContext);

  function getCredentials() {
    // TODO: THere has to be a better way to do what these two lines are doing.
    const usernameInput = document.querySelector(
      '[data-targetfield="username"] input',
    );
    const passwordInput = document.querySelector(
      '[data-targetfield="password"] input',
    );

    return { username: usernameInput.value, password: passwordInput.value };
  }

  function clearTimer() {
    if (timerHandle) {
      clearTimeout(timerHandle);
      timerHandle = null;
    }
  }

  function handleChange(event) {
    timerHandle = setTimeout(() => {
      clearTimer();

      const { username, password } = getCredentials();

      if (username.length > 0 && password.length > 0) {
        setSubmitDisabled(false);
      } else {
        setSubmitDisabled(true);
      }
    }, 250);
  }

  function clearAllFields(event) {
    clearTimer();

    document.querySelector('[data-targetfield="username"] input').value = "";
    document.querySelector('[data-targetfield="password"] input').value = "";
    setSubmitDisabled(true);
    setError(null);
  }

  async function doLogin(event) {
    const { username, password } = getCredentials();

    clearTimer();

    await login(username, password).catch((err) => {
      if (err.status === 401) {
        setError("Username or password is incorrect.");
      } else {
        setError(`${err.status}: ${err.message}`);
      }
    });

    const loginOkay = await loginStatus();
    setLoggedIn(loginOkay);
  }

  if (loggedIn) return null;
  if (!showLoginForm) return null;

  const modalRoot = document.getElementById("modal-root");

  return createPortal(
    <div className="login-container">
      <div className="login-form">
        <h1 className="login-header">Please Log In</h1>
        <div className="info">
          <OldLabeledInput
            modelName="login"
            fieldName="username"
            labelText="User Name"
            onChange={handleChange}
          />
          <OldLabeledInput
            modelName="login"
            fieldName="password"
            labelText="Password"
            type="password"
            onChange={handleChange}
          />
        </div>
        {error && <div className="error">{error}</div>}
        <div className="controls">
          <Button
            className="clear"
            disabled={false}
            onClick={clearAllFields}
            label="Clear All Fields"
          />
          <Button
            className="login"
            disabled={submitDisabled}
            onClick={doLogin}
            label="Login"
          />
        </div>
      </div>
    </div>,
    modalRoot,
  );
}

export default Login;
