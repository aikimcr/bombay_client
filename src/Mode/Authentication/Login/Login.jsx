import React, { useContext, useEffect, useRef, useState } from "react";
import { useForm } from "react-hook-form";
import BombayLoginContext from "../../../Context/BombayLoginContext";
import { loginStatus, login } from "../../../Network/Login";
import {
  Button,
  LabeledInput,
  PasswordInput,
  TextInput,
} from "../../../Components";
import { useRouteManager } from "../../../Hooks/useRouteManager";

import "./Login.scss";

export const Login = () => {
  const routeManager = useRouteManager();

  const { loggedIn, setLoggedIn } = useContext(BombayLoginContext);

  const [submitDisabled, setSubmitDisabled] = useState(true);

  const { register, reset, handleSubmit, formState, setError, clearErrors } =
    useForm({
      shouldUseNativeValidaton: true,
    });

  const doLogin = async (data) => {
    const { username, password } = data;

    try {
      await login(username, password);
    } catch (err) {
      if (err.status === 401) {
        setError("Username or password is incorrect.");
      } else {
        setError(`${err.status}: ${err.message}`);
      }
    }

    const loginOkay = await loginStatus();
    setLoggedIn(loginOkay);
  };

  const onSubmit = (data) => {
    doLogin(data);
  };

  const handleReset = () => {
    setSubmitDisabled(true);
    reset({
      username: "",
      password: "",
    });
  };

  useEffect(() => {
    if (formState.isValid && Object.entries(formState.errors).length === 0) {
      setSubmitDisabled(false);
      return;
    }

    setSubmitDisabled(true);
  }, [formState]);

  useEffect(() => {
    if (loggedIn) {
      routeManager.navigateToRoute("/artistList");
    }
  }, [loggedIn]);

  return (
    <div className="login-container">
      <form
        className="login-form"
        onSubmit={handleSubmit(onSubmit)}
        onReset={handleReset}
      >
        <h1 className="login-header">Please Log In</h1>
        <div className="info">
          <LabeledInput
            labelText="User Name"
            inputId="login-user-name"
            InputField={TextInput}
            inputProps={{
              ...register("username", {
                required: "Please enter a valid username",
              }),
              id: "login-user-name",
              name: "username",
            }}
          />
          <LabeledInput
            labelText="Password"
            inputId="login-password"
            InputField={PasswordInput}
            inputProps={{
              ...register("password", {
                required: "Please enter a valid password",
              }),
              id: "login-password",
              name: "password",
            }}
          />
        </div>
        {Object.entries(formState.errors).length !== 0 && (
          <div className="error">{formState.errors[0]}</div>
        )}
        <div className="controls">
          <Button
            className="clear"
            disabled={false}
            role="secondary"
            text="Clear All Fields"
            type="reset"
          />
          <Button
            className="login"
            disabled={submitDisabled}
            text="Login"
            type="submit"
          />
        </div>
      </form>
    </div>
  );
};
