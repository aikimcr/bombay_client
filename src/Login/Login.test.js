import * as NetworkLogin from '../Network/Login';
jest.mock('../Network/Login');

import { useState } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { act, fireEvent, render } from '@testing-library/react';

const testToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJGREM4MTEzOCIsInVzZXIiOnsiaWQiOjEsIm5hbWUiOiJhZG1pbiIsImFkbWluIjpmYWxzZX0sImlhdCI6MTY2NTk2NTA5OX0.2vz14X7Tm-oFlyOa7dcAF-5y5ympi_UlWyJNxO4xyS4';


import BombayLoginContext from '../Context/BombayLoginContext';
import BombayUtilityContext from '../Context/BombayUtilityContext';

import Login from './Login';

jest.useFakeTimers();

beforeEach(() => {
    localStorage.removeItem('jwttoken');

    const modalRoot = document.createElement('div');
    modalRoot.id = 'modal-root';
    document.body.append(modalRoot);
});

afterEach(() => {
    const modalRoot = document.getElementById('modal-root');
    modalRoot.remove();
});

it('should render the login mode', async () => {
    const { asFragment } = render(<ContextChanger loggedIn={false} showLoginForm={true}>
        <Login />
    </ContextChanger>);

    expect(asFragment).toMatchSnapshot();

    const modalRoot = document.getElementById('modal-root');
    expect(modalRoot).toMatchSnapshot();
});

it('should not render the login mode', async () => {
    render(<ContextChanger loggedIn={false} showLoginForm={false}>
        <Login />
    </ContextChanger>);

    const modalRoot = document.getElementById('modal-root');
    expect(modalRoot).toBeEmptyDOMElement();
});

it('should enable and disable the login button', async () => {
    render(<ContextChanger loggedIn={false} showLoginForm={true}>
        <Login />
    </ContextChanger>);

    const modalRoot = document.getElementById('modal-root');
    const loginButton = modalRoot.querySelector('.login.btn');
    expect(loginButton).toBeDisabled()

    const usernameInput = await changeInput(modalRoot, '[data-targetfield="username"] > input', 'herkimer', 250);
    expect(loginButton).toBeDisabled();

    const passwordInput = await changeInput(modalRoot, '[data-targetfield="password"] > input', 'jones', 250);
    expect(loginButton).not.toBeDisabled()
});

it('should login successfully', async () => {
    const { resolve } = NetworkLogin._setupMocks();

    const result = render(<ContextChanger loggedIn={false} showLoginForm={true}>
        <Login />
    </ContextChanger>);

    const modalRoot = document.getElementById('modal-root');
    const loginButton = modalRoot.querySelector('.login.btn');

    const usernameInput = await changeInput(modalRoot, '[data-targetfield="username"] > input', 'herkimer', 300);
    const passwordInput = await changeInput(modalRoot, '[data-targetfield="password"] > input', 'jones', 300);
    expect(loginButton).not.toBeDisabled()

    // Wait for the component to rerender.
    await act(async () => {
        const newLoginButton = modalRoot.querySelector('.login.btn');
        newLoginButton.click();
    });

    await act(async () => {
        resolve(testToken);
    });

    expect(NetworkLogin.login).toBeCalledTimes(1);
    expect(NetworkLogin.login).toBeCalledWith('herkimer', 'jones');
});

it('should show error on failed login', async () => {
    const { resolve } = NetworkLogin._setupMocks();

    const result = render(<ContextChanger loggedIn={false} showLoginForm={true}>
        <Login />
    </ContextChanger>);

    const modalRoot = document.getElementById('modal-root');
    const loginButton = modalRoot.querySelector('.login.btn');

    const usernameInput = await changeInput(modalRoot, '[data-targetfield="username"] > input', 'herkimer', 250);
    const passwordInput = await changeInput(modalRoot, '[data-targetfield="password"] > input', 'jones', 250);
    expect(loginButton).not.toBeDisabled()

    // Wait for the component to rerender.
    await act(async () => {
        const newLoginButton = modalRoot.querySelector('.login.btn');
        newLoginButton.click();
    });

    await act(async () => {
        resolve(null);
    });

    expect(NetworkLogin.login).toBeCalledTimes(1);
    expect(NetworkLogin.login).toBeCalledWith('herkimer', 'jones');
});

it('should show error on call error', async () => {
    const { resolve } = NetworkLogin._setupMocks();

    const result = render(<ContextChanger loggedIn={false} showLoginForm={true}>
        <Login />
    </ContextChanger>);

    const modalRoot = document.getElementById('modal-root');
    const loginButton = modalRoot.querySelector('.login.btn');

    const usernameInput = await changeInput(modalRoot, '[data-targetfield="username"] > input', 'herkimer', 250);
    const passwordInput = await changeInput(modalRoot, '[data-targetfield="password"] > input', 'jones', 250);
    expect(loginButton).not.toBeDisabled()

    // Wait for the component to rerender.
    await act(async () => {
        const newLoginButton = modalRoot.querySelector('.login.btn');
        newLoginButton.click();
    });

    await act(async () => {
        resolve(null);
    });

    expect(NetworkLogin.login).toBeCalledTimes(1);
    expect(NetworkLogin.login).toBeCalledWith('herkimer', 'jones');
});

it('should clear the inputs and error on clear all fields', async () => {
    const { resolve } = NetworkLogin._setupMocks();

    const result = render(<ContextChanger loggedIn={false} showLoginForm={true}>
        <Login />
    </ContextChanger>);

    const modalRoot = document.getElementById('modal-root');
    const loginButton = modalRoot.querySelector('.login.btn');

    const usernameInput = await changeInput(modalRoot, '[data-targetfield="username"] > input', 'herkimer', 250);
    const passwordInput = await changeInput(modalRoot, '[data-targetfield="password"] > input', 'jones', 250);
    expect(loginButton).not.toBeDisabled()

    // Wait for the component to rerender.
    await act(async () => {
        const newLoginButton = modalRoot.querySelector('.login.btn');
        newLoginButton.click();
    });

    await act(async () => {
        resolve(null);
    });

    const clearButton = modalRoot.querySelector('.clear.btn');

    // Wait for the component to rerender.
    await act(async () => {
        clearButton.click();
    });

    expect(loginButton).toBeDisabled()
});