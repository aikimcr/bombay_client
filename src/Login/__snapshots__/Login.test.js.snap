// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the login mode 1`] = `[Function]`;

exports[`should render the login mode 2`] = `
<div
  id="modal-root"
>
  <div
    class="login-container"
  >
    <div
      class="login-form"
    >
      <h1
        class="login-header"
      >
        Please Log In
      </h1>
      <div
        class="info"
      >
        <div
          class="labeled-input"
          data-fieldname="username"
          data-modelname="login"
        >
          <label
            for="login-username-isNew"
          >
            User Name
          </label>
          <input
            id="login-username-isNew"
            name="name"
            type="text"
            value=""
          />
        </div>
        <div
          class="labeled-input"
          data-fieldname="password"
          data-modelname="login"
        >
          <label
            for="login-password-isNew"
          >
            Password
          </label>
          <input
            id="login-password-isNew"
            name="name"
            type="password"
            value=""
          />
          <input
            class="toggle"
            data-hidden="true"
            type="button"
            value="Show"
          />
        </div>
      </div>
      <div
        class="controls"
      >
        <button
          class="clear btn"
        >
          Clear All Field
        </button>
        <button
          class="login btn"
          disabled=""
        >
           
          Login
        </button>
      </div>
    </div>
  </div>
</div>
`;
